Refactor the reference section of the docs

tybug opened the pull request:
Still a large pr but there's no sane way to split most of this without causing more work for me, thanks to all the inter-references. I could split the redirect extension off is probably the biggest thing

There's a minor amount of added or removed content here, but most of it is moving around existing stuff into their own page

tybug commented on hypothesis-python/docs/reference/django.rst:
```diff
@@ -1,11 +1,11 @@
-=====================
-Hypothesis for Django
-=====================
+======
+Django
+======
```
I could be talked into keeping "Hypothesis for Django". I lean pretty minimalist in general (in life, designs, wording, etc), but I realize I can go too far sometimes.

tybug commented on hypothesis-python/docs/reference/reproducing-failures.rst:
```diff
@@ -1,19 +1,8 @@
-====================
 Reproducing failures
 ====================
```
My heading ordering is single =, single -, single ~ (and if you need to go beyond that you're probably doomed anyway). I try to avoid double headers for anything, but I think I might be the odd one out here, because e.g. the python rst guide recommends double headers for chapters. In defense of single headers, the python docs have *way* more structure than we do!

tybug commented on hypothesis-python/docs/reference/strategies.rst:
```diff
@@ -0,0 +1,120 @@
+.. autofunction:: hypothesis.strategies.runner
+.. autofunction:: hypothesis.strategies.shared
+
+Misc
+----
+
+.. autofunction:: hypothesis.strategies.builds
+.. autofunction:: hypothesis.strategies.functions
+.. autofunction:: hypothesis.strategies.slices
+
+.. autofunction:: hypothesis.strategies.sampled_from
+.. autofunction:: hypothesis.strategies.permutations
+
```
categorization is hard...would be happy to hear advice on how to categorize the strategies (either these misc ones, or any of the other ones too)

jobh commented on hypothesis-python/docs/reference/targeted.rst:
```diff
@@ -0,0 +1,23 @@
+      ab_c = (a + b) + c
+      a_bc = a + (b + c)
+      difference = abs(ab_c - a_bc)
+      target(difference)  # Without this, the test almost always passes
```
As fas as I can tell, `target` makes little difference here. Around 40% discover rate with or without it.

tybug replied:
same..I'm not going to follow up on this right now, but I've added it as a TODO to https://github.com/HypothesisWorks/hypothesis/issues/4309

jobh replied:
Thanks! A comment for that time (ignore for this PR) is that `target` likely doesn't work well here because the difference is constant for large regions of the input space.

We might want to comment that the target should be fairly smooth, not discrete-valued, and not added willy-nilly as it may actually reduce bug-finding power with the default example budget (as indicated below).

For background, the optimiser seems to get more stuck than I'd expect in these flat regions. E.g.,
```
@given(st.floats())
@settings(database=None, phases=[Phase.generate, Phase.target])
def test_target(f):
    target(f > 1e5)
    print(f)
```
doesn't vary `f` very much inside the `True` region. Fair enough, but worth noting.

Zac-HD commented:
Quick thoughts again:
- Yeah, this is going to be about at our size limit, but I agree that chopping it up more wouldn't really be a net improvement.
- IMO, searchability is really important for API reference docs.  "land on page, cmd-f, done" is a key usecase that e.g. [pytest](https://docs.pytest.org/en/stable/reference/reference.html) serves well these days.
  - If that's way too large, one page for all of the strategies, and one for everything else?
  - Some of the stuff currently in reference docs will presumably move to explanations or how-tos later.
- Makes sense to switch from `.. automodule:: ...`, but we should then add a test that we document every function in the strategies module (or perhaps everything listed in any `__all__` attribute?
- Haven't run through all the detail yet, since I'm expecting files to move around again.

tybug commented:
yeah it's a fair point, I'm on board with one big page for the api reference. It'll be a bit awkward until we move the more tutorial-style things out from the api reference, but still an improvement on the status quo.

tybug commented:
so I know we said just one or two reference pages, but I was working through this and I think there's a valid distinction between "api/coding reference" and "'meta' reference", like observability or the pytest plugin. They don't feel right to me in an api reference page, because they don't have an in-library method exposed by Hypothesis.

tybug commented on hypothesis-python/docs/typing.rst:
I think we will eventually want to move this page somewhere better ("how to write type hints for your strategies"?), or remove it entirely, but top level is fine during the transition

tybug commented:
I've added a test to check that all exported names are documented, as recommended. This turned up a bug in the current docs where `hypothesis.extra.numpy.boolean_dtypes` is not documented. I couldn't figure out why, and since it's not a new issue I've added it as a todo item in the tracking list.

Zac-HD commented on hypothesis-python/docs/reference/strategies.rst:
```diff
@@ -1,15 +1,216 @@
+.. _strategies:
+
+Strategies reference
+====================
+
+Primitives
+----------
```
I think we'd really benefit from a paragraph of explanation right at the top of the reference page: what is a strategy?  the fact that you can combine them in various ways, and a link to `@given()`.

Zac-HD commented on hypothesis-python/docs/reference/strategies.rst:
```diff
@@ -1,15 +1,216 @@
+.. autofunction:: hypothesis.strategies.just
+.. autofunction:: hypothesis.strategies.booleans
+
+Numeric
+-------
+
+.. autofunction:: hypothesis.strategies.integers
```
add a top-of-section (italic, not admonition) _See also: ref-numpy, ref-arrays, ref-pandas_.

Zac-HD commented on hypothesis-python/docs/reference/strategies.rst:
```diff
@@ -1,15 +1,216 @@
+.. autofunction:: hypothesis.strategies.decimals
+.. autofunction:: hypothesis.strategies.fractions
+
+Text
```
"Strings", since we're including `binary()` here (which I think is the right place)

Zac-HD commented on hypothesis-python/docs/reference/strategies.rst:
```diff
@@ -1,15 +1,216 @@
+.. autofunction:: hypothesis.strategies.binary
+.. autofunction:: hypothesis.strategies.emails
+.. autofunction:: hypothesis.strategies.uuids
+.. autofunction:: hypothesis.strategies.ip_addresses
```
These aren't strings, they're [`IPAddress` objects](https://docs.python.org/3/library/ipaddress.html), and thus probably belong in misc.

tybug replied:
same with UUIDS then?

Zac-HD replied:
🤦‍♂️ yes.  We could add a _see also_ note at the top of the section though, noting that you can access uuid / ip address strings via `st.uuids(...).map(str)` etc.

Zac-HD commented on hypothesis-python/docs/reference/strategies.rst:
```diff
@@ -1,15 +1,216 @@
+Misc
+----
+
+.. autofunction:: hypothesis.strategies.builds
```
belongs up in combinators imo.

Zac-HD commented on hypothesis-python/docs/reference/strategies.rst:
```diff
@@ -1,15 +1,216 @@
+Hypothesis offers strategies for `Array API <https://data-apis.org/>`_ adopting
+libraries in the ``hypothesis.extra.array_api`` package. See :issue:`3037` for
+more details.  If you want to test with :pypi:`CuPy <cupy>`, :pypi:`Dask <dask>`, :pypi:`JAX <jax>`,
+:pypi:`MXNet <maxnet>`, :pypi:`PyTorch <torch>`, :pypi:`TensorFlow <tensorflow>`, or :pypi:`Xarray <xarray>` -
+or just :pypi:`NumPy <numpy>` - this is the extension for you!
```
With [Xarray shipping upstream strategies](https://docs.xarray.dev/en/stable/user-guide/testing.html) (see https://github.com/pydata/xarray/issues/6911 and linked PRs), we might want to link to those docs + rephrase to something like "or TensorFlow - this is the extension for you!  Of course it also works with libraries like numpy-ref or xarray-link which have more specific strategies, if you want to stick to the standard array API.

tybug replied:
went for a slightly different format with a .. note

Zac-HD commented on hypothesis-python/docs/reference/strategies.rst:
```diff
@@ -1,15 +1,216 @@
+:func:`~xps.arrays` and :func:`~xps.from_dtype`, but based on the Array API
+standard semantics and returning objects from the ``xp`` module:
+
+.. automodule:: xps
+   :members:
+        from_dtype,
```
Oooh... does your excellent new test check that this is complete?  It's not an importable module, exactly...

Zac-HD reviewed (commented):
More scattered thoughts, and appreciation for the ongoing work 🙏

tybug commented on hypothesis-python/docs/reference/index.rst:
```diff
@@ -0,0 +1,13 @@
+.. toctree::
+   :maxdepth: 1
+
+   strategies
+   api
+   ghostwriter
+   Pytest plugin <pytest-plugin>
+   observability
```
hmm, I'm on board with everything, except dropping the index page. That would make the "Reference" header on the left sidebar not a hyperlink, and means we have nowhere to link to on the eventual frontpage diataxis square, in the "Reference" corner.

Zac-HD replied:
Ah, fair enough - that's two good reasons to keep it!

Once we've got an index page though, that's a good place to put a short description of what's on each subpage (paragraph total?), not just the bare links.

tybug replied:
yeah we can expand the content on that page for sure

tybug commented on hypothesis-python/docs/conf.py:
```diff
@@ -40,10 +40,33 @@
+redirects = {
+    "details": "reference/index.html",
+    "data": "reference/strategies.html",
+    "database": "reference/api.html",
+    "stateful": "reference/api.html",
+    "reproducing": "reference/api.html",
```
oh we definitely need to do this. Will require changes to the redirect extension, but I view this as a requirement before merge

tybug commented on hypothesis-python/docs/reference/control.rst:
```diff
@@ -0,0 +1,42 @@
+Control
+=======
```
agree, but it's a nice categorization name. I'm not sure where else to put them or what other name to give this seciton

Zac-HD replied:
Put `target()` in the same group, I guess it might as well be called Control for now.  (note & event are more like observability but 🤷‍♂️)

Overall I'd order the api-reference page given, example, control (?), setttings, database, seed & reproduce (cross-refs example), fuzzing.  It should also get a lot easier to understand as we move stuff out to how-to pages and maybe some explanations 😅

tybug commented on hypothesis-python/docs/reference/pytest-plugin.rst:
```diff
@@ -0,0 +1,74 @@
+Test statistics
+---------------
+
+You can see a number of statistics about executed tests by passing the command line argument ``--hypothesis-show-statistics``. This will include some general statistics about the test:
```
we should link really prominently to tyche at the top of "observability" as well? Otherwise end-users landing at observability don't know what to do; they're not going to go and implement their own observability ui!

Zac-HD commented on hypothesis-python/docs/reference/index.rst:
```diff
@@ -0,0 +1,11 @@
+.. toctree::
+   :maxdepth: 1
+
+   strategies
+   api
```
reverse these?

tybug replied:
I'd guess that most people who look up the hypothesis docs are looking up a strategy! I guess the ordering is a bit awkward with the sidebar heading being "API Reference" and then its nested "API Reference" page being second, not first.

Zac-HD commented on hypothesis-python/docs/reference/integrations.rst:
```diff
@@ -0,0 +1,243 @@
+
+.. note::
+
+  :ref:`Observability <observability>` tools offers
```
incomplete

Zac-HD commented on hypothesis-python/docs/reference/integrations.rst:
```diff
@@ -0,0 +1,243 @@
+Integrations Reference
+======================
+
+Reference for Hypothesis features with a defined interface, but no code API.
```
IMO makes sense to move the CLI extra docs here, including codemods.

tybug replied:
hmm, are you imagining that we would remove the `extras.rst` page and distribute the extras docs among other pages as appropriate? The "home" for each extra isn't the same, so cli/codemods I agree goes in integrations, but lark/pytz/dateutil would go in strategies. We could do this, and have a single small section which lists all of the available extras and links out to the appropriate references?

Zac-HD replied:
Yeah, I think that's the way to do it - especially if we want cmd-f searchability of the reference docs, which I do think is valuable.

tybug replied:
done

Zac-HD commented on hypothesis-python/docs/reference/strategies.rst:
```diff
@@ -0,0 +1,230 @@
+Strategies Reference
+====================
+
+Strategy objects tell Hypothesis what types of inputs to generate. For instance, passing ``st.lists(st.integers(), min_size=1)`` to |@given| tells Hypothesis to generate lists of integers with at least one element.
+
+Strategies can be combined using :ref:`combinator strategies <combinators>`, or modified using |strategy.filter|, |strategy.map|, or |strategy.flatmap|.
```
```suggestion
Strategies are the way we describe values for |@given| to generate.  For instance, ``st.lists(st.integers(), min_size=1)`` tells Hypothesis to generate lists of integers with at least one element.

This reference page lists all of Hypothesis' first-party functions which return a strategy; there are also many provided by <third-party libraries>(xref), and you can define your own too.  Note that people often say "strategy" when they mean "function returning a strategy"; it's usually clear from context which they meant.

Strategies can be passed to many strategy functions as arguments, combined using :ref:`combinator strategies <combinators>`, and modified using |strategy.filter|, |strategy.map|, or |strategy.flatmap|.
```

tybug replied:
> Strategies can be passed to many strategy functions as arguments

I get the distinction you're making, and I think it's worth making, but this is a confusing sentence to parse imo. Bite the technical incorrectness and say "strategies can be passed to other strategies as arguments"?

I've applied a variant of this, but happy to tweak further

Zac-HD replied:
I do think we need to call out both the distinction and the common sloppy usage in this explanatory header, but don't care exactly how we do that.

Zac-HD commented on hypothesis-python/docs/reference/strategies.rst:
```diff
@@ -0,0 +1,230 @@
+.. autofunction:: hypothesis.strategies.sampled_from
+.. autofunction:: hypothesis.strategies.permutations
+
+Provisional
+-----------
+
+.. automodule:: hypothesis.provisional
+  :members:
+  :exclude-members: DomainNameStrategy
```
We've only got two strategies in here; I'd be inclined to put them under `Strings` above.  (and include provisional in our all-documented test)

tybug replied:
I'm on board with this. How strongly should we be `.. warning` these strategies? I've never actually understood why these are provisional, the interface for domains and urls seem stable enough to me.

Zac-HD replied:
Not strongly imo, they're provisional more because of iirc unresolved design questions than real instability.

Zac-HD commented on hypothesis-python/docs/reference/strategies.rst:
```diff
@@ -0,0 +1,230 @@
+.. autoclass:: hypothesis.strategies.DataObject
+.. autoclass:: hypothesis.strategies.DrawFn
+
+.. autoclass:: hypothesis.strategies.SearchStrategy
+
+  .. automethod:: hypothesis.strategies.SearchStrategy.filter
+  .. automethod:: hypothesis.strategies.SearchStrategy.map
+  .. automethod:: hypothesis.strategies.SearchStrategy.flatmap
```
I think we can use syntax `:members: filter, map, flatmap` for this?

tybug replied:
this didn't work initially and I gave up trying to figure out why /shrug. It might be because we're overriding `__module__: str = "hypothesis.strategies"` in `SearchStrategy`, which is a pretty cursed thing to be doing in general tbh.

Zac-HD commented on hypothesis-python/docs/changes.rst:
```diff
@@ -959,7 +959,7 @@ in :func:`~hypothesis.strategies.from_type`.
 --------------------
 
-Add better error message for :obj:`~python:typing.TypeIs` types
+Add better error message for :obj:`!~python:typing.TypeIs` types
```
regression due to merge conflict I think?

Zac-HD commented on hypothesis-python/docs/reference/api.rst:
```diff
@@ -0,0 +1,1038 @@
+
+   Following the first failure, Hypothesis will (:ref:`usually <phases>`) track which
+   lines of code are always run on failing but never on passing inputs.
+   This relies on :func:`python:sys.settrace`, and is therefore automatically disabled on
```
or sys.monitoring, on newer versions of Python!

Zac-HD commented on hypothesis-python/docs/reference/api.rst:
```diff
@@ -0,0 +1,1038 @@
+Default settings
+~~~~~~~~~~~~~~~~
+
+At any given point in your program there is a current default settings,
+available as ``settings.default``. As well as being a settings object in its own
+right, all newly created settings objects which are not explicitly based off
+another settings are based off the default, so will inherit any values that are
+not explicitly set from it.
```
iirc this is actually wrong: `settings.default` is the _default settings_, which are immutable, and you can access the _current settings_ by constructing a new object with `settings()`.  [whoops 😓]

tybug replied:
they're actually both equivalent! settings.default gets you the current settings as well. We should recommend one and only one correct way to do this though. I'd vote for `settings()`, because `settings.default` implies, well, the default settings, not the current settings.

Zac-HD commented on hypothesis-python/docs/reference/api.rst:
```diff
@@ -0,0 +1,1038 @@
+.. autoclass:: hypothesis.database.BackgroundWriteDatabase
+.. autoclass:: hypothesis.extra.redis.RedisExampleDatabase
+
+Defining your own ExampleDatabase
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+You can define your :class:`~hypothesis.database.ExampleDatabase`, for example
+to use a shared datastore, with just a few methods:
+
+.. autoclass:: hypothesis.database.ExampleDatabase
+   :members:
+
+
+.. code:: python
+
+    @given(x=integers())
+    @pytest.mark.trio
+    async def test(x): ...
+
+
+    # Illustrative code, inside the pytest-trio plugin
+    test.hypothesis.inner_test = lambda x: trio.run(test, x)
+
+For authors of test runners however, assigning to the ``inner_test`` attribute of the ``hypothesis`` attribute of the test will replace the interior test.
```
We seem to have lost some content here / should move the trio stuff to bottom of doc.

tybug replied:
this all belongs in `Custom function execution`, whoops

Zac-HD commented on hypothesis-python/docs/reference/api.rst:
```diff
@@ -0,0 +1,1038 @@
+django test suite (this is important because your test function will be called
+multiple times and you don't want them to interfere with each other). Test cases
+on these classes that do not use
+:func:`@given <hypothesis.given>` will be run as normal.
```
"as normal for <xref Django TestCase cls> or <ditto TransactionTestCase>."

tybug replied:
I'm not following what you're suggesting here sorry!

Zac-HD replied:
Oh wow, markdown rendering issue!  Added backticks to clarify.

tybug replied:
wow, our django links go to a documentation site that I didn't even know existed!! https://django.readthedocs.io/en/stable/index.html#. I've updated this to the actual django docs at eg https://docs.djangoproject.com/en/5.1/topics/testing/tools/#django.test.TransactionTestCase

Zac-HD commented on hypothesis-python/docs/reference/api.rst:
```diff
@@ -0,0 +1,1038 @@
+
+We recommend avoiding :class:`~hypothesis.extra.django.TransactionTestCase`
+unless you really have to run each test case in a database transaction.
+Because Hypothesis runs this in a loop, the performance problems it normally has
```
it -> TransactionTestCase, to avoid implying that Hypothesis has perf problems.

Zac-HD commented on hypothesis-python/docs/reference/api.rst:
```diff
@@ -0,0 +1,1038 @@
+For the moment there's no explicit support in hypothesis-django for generating
+dependent models. i.e. a Company model will generate no Shops. However if you
+want to generate some dependent models as well, you can emulate this by using
+the *flatmap* function as follows:
```
xref flatmap _method_.

Zac-HD commented on hypothesis-python/docs/reference/api.rst:
```diff
@@ -0,0 +1,1038 @@
+The way flatmap works is that we draw a value from the original strategy, then
+apply a function to it which gives us a new strategy. We then draw a value from
+*that* strategy. So in this case we're first drawing a company, and then we're
+drawing a list of shops belonging to that company: The *just* strategy is a
```
xref `just`

Zac-HD reviewed (approved):
I think we're done with this part - obviously there are plenty of followups, but I feel good about shipping this as a single cohesive change for our users 😁
